[FLINK-8777][state]Improve resource release for local recovery

sihuazhou opened the pull request:
## What is the purpose of the change

This PR fixes [FLINK-8777](https://issues.apache.org/jira/browse/FLINK-8777). When recovery from failed, `TaskLocalStateStoreImpl.retrieveLocalState()` will be invoked, we can release all entry from `storedTaskStateByCheckpointID`  that does not satisfy `entry.checkpointID == checkpointID`, this can prevent the resource leak when job loop in `local checkpoint completed => failed => local checkpoint completed => failed ...`.

## Brief change log

  - release resource in retrieveLocalState
  - change the type of toDiscard from Map to a single entry.

## Verifying this change

This changes can be verified by the exists tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

sihuazhou commented:
@StefanRRichter Could you please have a look at this?

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -126,30 +133,54 @@ public void storeLocalState(
+
+			if (retrieveWithDiscard) {
+				// Only the TaskStateSnapshot.checkpointID == checkpointID is useful, we remove the others
+				final List<Map.Entry<Long, TaskStateSnapshot>> toRemove = new ArrayList<>();
```
I think we can de-duplicate code by extracting a method with the code from `confirmCheckpoint(...)`, maybe called `pruneCheckpoints(...)`. We can do the comparison for both use-cases as `entryCheckpointId != checkpointID` and have a boolean parameter which determines if we break the iteration in the `else` case or not.

sihuazhou replied:
👍 addressed.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -90,6 +92,10 @@
 	private boolean disposed;
 
+	/** Whether to discard the useless state when retrieve local checkpoint state. */
+	@Nonnull
```
addressed.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -126,30 +133,54 @@ public void storeLocalState(
 		synchronized (lock) {
 			TaskStateSnapshot snapshot = storedTaskStateByCheckpointID.get(checkpointID);
+
+			Iterator<Map.Entry<Long, TaskStateSnapshot>> entryIterator =
```
addressed.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -285,4 +316,9 @@ public String toString() {
 	}
+
+	@VisibleForTesting
+	void setRetrieveWithDiscard(@Nonnull boolean retrieveWithDiscard) {
```
addressed.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -90,6 +92,10 @@
 
+	/** Whether to discard the useless state when retrieve local checkpoint state. */
+	@Nonnull
+	private boolean retrieveWithDiscard;
```
addressed.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -300,6 +291,32 @@ private void deleteDirectory(File directory) throws IOException {
+	 * Pruning the useless checkpoints.
+	 */
+	private void pruneCheckpoints(long checkpointID, boolean breakTheIteration) {
+
```
I suggest to add an assert that the thread holds `lock` and document that this method should be called only when holding the lock.

sihuazhou replied:
👍

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -159,6 +166,11 @@ public TaskStateSnapshot retrieveLocalState(long checkpointID) {
 			snapshot = storedTaskStateByCheckpointID.get(checkpointID);
+
+			if (retrieveWithDiscard) {
+				// Only the TaskStateSnapshot.checkpointID == checkpointID is useful, we remove the others
```
Comment is no longer required.

sihuazhou replied:
👍

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -90,6 +92,9 @@
 	private boolean disposed;
 
+	/** Whether to discard the useless state when retrieve local checkpoint state. */
+	private boolean retrieveWithDiscard = true;
```
Why do we need this? Is there any case for not doing the cleanup?

sihuazhou replied:
Aha, this is just for passing the existing test case in `TaskLocalStateStoreImplTest` ... 
```java
	private void checkStoredAsExpected(List<TaskStateSnapshot> history, int off, int len) throws Exception {
		for (int i = off; i < len; ++i) {
			TaskStateSnapshot expected = history.get(i);
			Assert.assertTrue(expected == taskLocalStateStore.retrieveLocalState(i));
			Mockito.verify(expected, Mockito.never()).discardState();
		}
	}
```

StefanRRichter replied:
Then there are two better options in my opinion, because the flag is pure boilerplate:

- Change the test to check what we are doing now, because that is what happens in the real use-case.
- Maybe even better: split the method `retrieveLocalState` further: one method for pruning, one package-private method that does all the pure retrieval, logging, and `null` transformation. In the old `retrieveLocalState`, do the cleanup first, then the pure retrieval/logging. Call the package private method in the test.

Maybe the test should then also just do both?

sihuazhou replied:
Agreed and I prefer the second option.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -300,6 +291,32 @@ private void deleteDirectory(File directory) throws IOException {
+			Map.Entry<Long, TaskStateSnapshot> snapshotEntry = entryIterator.next();
+			long entryCheckpointId = snapshotEntry.getKey();
+
+			if (entryCheckpointId != checkpointID) {
```
After a second though, while I think this code is currently correct, the case with breaking looks a bit dangerous. Potentially, if the checkpoint id is not there, this would not stop and prune ongoing checkpoints. I wonder if we should make the `if` a bit more complex, but safer (checking that the breaking case never exceeds the checkpoint id). What do you think?

sihuazhou replied:
I agree with you that the breaking case looks a bit dangerous ... I think maybe we could pass a `Predicate` for the `if` and let the caller side pass the `Predicate` into this function. This could make it cleaner from the caller side and don't need to mass the logic into the `if` to make it complex.

StefanRRichter replied:
That is fine, from my point of view that is just one way of making the `if` more complex.

sihuazhou commented:
@StefanRRichter I have addressed your suggestions, except the one that to make the `if` a bit complex, instead I introduced a `Predicate` for the `pruneCheckpoints()`. I not sure whether it is ok to you, if you still against doing so, I'd like to change the code as to make the `if` a bit complex.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java:
```diff
@@ -300,6 +295,34 @@ private void deleteDirectory(File directory) throws IOException {
+	/**
+	 * Pruning the useless checkpoints, it should be called only when holding the {@link #lock}.
+	 */
+	private void pruneCheckpoints(Predicate<Long> pruningChecker, boolean breakOnceCheckerFalse) {
```
We can use `LongPredicate` instead of `Predicate<Long>`

sihuazhou replied:
👍 I prefer to use `LongPredicate`, addressing ...

StefanRRichter commented:
Thanks, I will merge this once the final points are addressed :-)

StefanRRichter commented:
Ok, I took another look at the complete picture and from the test got the feeling that retrieval and pruning should be two separated concerns and that not only should we have two internal methods, but maybe also expose them as different methods. For the sake to keep this short, I made a proposal in this branch:

https://github.com/StefanRRichter/flink/tree/improve_resource_release_for_local_recovery

If you like the change, I would squash it and commit under your name, because you did all of the important parts. What do you think?

sihuazhou commented:
@StefanRRichter Thanks, I like it! It looks very good.
